raiden_contracts/contracts/MonitoringService.sol

```diff
@@ -0,0 +1,321 @@
+
+        // Only allow PBs with higher nonce to be submitted
+        require(reward.nonce < nonce);
+
+        // MSC stores channel_identifier, MS_address, reward_amount, nonce
+        // of the MS that provided the balance_proof with highest nonce
+        rewards[channel_identifier] = Reward({
+            reward_proof_signature: reward_proof_signature,
+            nonce: nonce,
+            reward_sender_address: reward_sender_address
+        });
+	}
+
+    /// @notice Called by a registered MS, when providing a new balance proof
+    /// to a monitored channel.
+    /// Can be called multiple times by different registered MSs as long as the PB provided
+    /// is newer than the current newest registered BP.
+    /// @param nonce Strictly monotonic value used to order BPs
+    /// omitting PB specific params, since these will not be provided in the future
+    /// @param reward_proof_signature Signature of the Raiden Node signing the reward
+    /// @param token_network_address Address of the Token Network in which the channel
+    /// being monitored exists.
+    function monitor(
+        bytes32 balance_hash,
+        uint256 nonce,
+        bytes32 additional_hash,
+        bytes closing_signature,
+        bytes non_closing_signature,
+        address reward_sender_address,
+        bytes reward_proof_signature,
+        address token_network_address,
+        address closing_participant,
+        address non_closing_participant
+    )
+        isMonitor(msg.sender)
+        public
+    {
+        updateReward(
+            token_network_address,
+            closing_participant,
+            non_closing_participant,
+            reward_proof_signature,
+            nonce,
+            reward_sender_address
+        );
+        TokenNetwork token_network = TokenNetwork(token_network_address);
+
+
+        // Call updateTransfer in the corresponding TokenNetwork
+        token_network.updateNonClosingBalanceProof(
+            closing_participant,
+            non_closing_participant,
+            balance_hash,
+            nonce,
+            additional_hash,
+            closing_signature,
+            non_closing_signature
+        );
+
+        emit NewBalanceProofReceived(
+            reward_proof_signature,
+            nonce,
+            msg.sender,
+            reward_sender_address
+        );
+    }
+
+    /// @notice Called after a monitored channel is settled in order for MS to claim the reward
+    /// Can be called once per settled channel by everyone on behalf of MS
+    /// @param token_network_address Address of the Token Network in which the channel
+    /// @param reward_amount Amount to be paid as reward to the MS
+    /// @param monitor_address Address of the MS
+    function claimReward(
+        address token_network_address,
+        uint192 reward_amount,
+        address monitor_address,
+        address closing_participant,
+        address non_closing_participant
```

Comment:
`claimReward` should only need the `balance_proof_id`. The contract will know who to send the tokens to.